src/elf.rs

```diff
@@ -731,27 +780,46 @@ impl<E: UserDefinedError, I: InstructionMeter> Executable<E, I> {
             ro_slices.push((section_addr, section_data));
         }
 
-        if highest_addr > elf_bytes.len()
-            || (config.reject_broken_elfs
-                && lowest_addr.saturating_add(ro_fill_length) > highest_addr)
-        {
+        if config.reject_broken_elfs && lowest_addr.saturating_add(ro_fill_length) > highest_addr {
             return Err(ElfError::ValueOutOfBounds);
         }
 
-        let can_borrow = !have_offsets
+        let can_borrow = !invalid_offsets
             && last_ro_section
                 .saturating_add(1)
                 .saturating_sub(first_ro_section)
                 == n_ro_sections;
         let ro_section = if config.optimize_rodata && can_borrow {
             // Read only sections are grouped together with no intermixed non-ro
             // sections. We can borrow.
-            Section::Borrowed(lowest_addr..highest_addr as usize)
+
+            // Section addresses and their corresponding byte offsets are
+            // translated by a constant offset. Subtract the offset to get byte
+            // positions.
+            let lowest_byte = lowest_addr.saturating_sub(addr_byte_offset.unwrap_or(0) as usize);
+            let highest_byte = highest_addr.saturating_sub(addr_byte_offset.unwrap_or(0) as usize);
```

Comment:
This is not really the last byte but the one behind it, right?
Might be better to call these `start` and `end` instead.

Also the distinction between `byte` and `addr` is not strong enough IMO. How about virtual address, file offset, etc. ?

Reply:
Ack to both. I didn't want to use *file* offset because it's technically a buffer, but goblin calls it `file_range()` too so I'll change it to that for consistency.

Reply:
I've renamed the two variables above to `buf_offset_start` and `buf_offset_end`, and updated the rest of the code to use the `buf` prefix when dealing with vecs/slices. So now `addr_`  is used for virtual addresses, `buf_` is used for  vecs and slices. Eg:

```
           Section::Borrowed(addr_offset, buf_offset_start..buf_offset_end as usize)
...
           let mut ro_section = vec![0; buf_len];
            for (section_addr, slice) in ro_slices.iter() {
                let buf_offset_start = section_addr.saturating_sub(lowest_addr);
                ro_section[buf_offset_start..buf_offset_start.saturating_add(slice.len())]
                    .copy_from_slice(slice);
            }
```

I personally prefer this to "file", but I don't really feel strongly about this so if you do, I'm happy to change buf to file.